### ISSUE TITLE
Partial aliasing in place operations should warn

### ISSUE THREAD

**Issue by davidbau** [pytorch/pytorch#8212]
Sometimes an in-place division where the denominator comes from the same tensor will yield unpredictable results.  For example, if you want an array that counts from 0 to 100 in each of 50 rows, but actually you want to normalize to the range [0..1], it seems natural to write:

```
import torch
c = torch.arange(101)[None,:].repeat(50, 1)
c /= c[:,100:]
print(c.max())
```

However, running this on CPU repeatedly will produce an unpredictable max, e.g., 82, 95, 87, etc, rather than the expected 1.0.

This operation appears to be reliable on numpy, and for many array shapes it also works fine in pytorch.  But it fails on certain sizes, leading to oddly rare and difficult-to-find bugs.

Should `c /= c[:,-1:]` be reliable?  If not, would it be reasonable to warn on such operations?

**Comment by fmassa** [comment]
This is a duplicate of https://github.com/pytorch/pytorch/issues/906 and I think won't be fixed anytime soon.

**Comment by davidbau** [comment]
Makes sense.  Would a warning be feasible?  Even knowing that this is an issue, it is easy to make this mistake by accident.

**Comment by apaszke** [comment]
Mentioning diophantine equations sounds as if it was a computationally hard problem, but I think that since the problem that arises is in fact a single linear equation with `2d` variables (`d` is the number of dimensions), then we could actually implement the aliasing check relatively efficiently, and it would be exact. A simplified analysis leads me to conclude that it can be done in approx `c * 64 * d` arithmetic operations where `c` is a small constant like 2-4. On the other hand, this means it might need a microsecond or so, and that is a long time for us, right? 😄

**Comment by davidbau** [comment]
Cool.  Numpy's conservative diophantine solver seems to check the reasonable cases quickly.  Maybe could be used.

https://github.com/numpy/numpy/blob/master/numpy/core/src/private/mem_overlap.c#L756

They use this before every array assignment that looks like it might alias, for example

https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/array_assign_array.c#L297

**Comment by apaszke** [comment]
I only took a cursory look at it, but I think that the proof of the [One-variable linear case from Wikipedia](https://en.wikipedia.org/wiki/Diophantine_equation#One_equation) could be generalized to multivariate case. If this is the case, then you can just use the euclidean algorithm to find any solution, and try to adjust dimension indices independently (they could have been negative or large) to make them fall into the desired range.

**Comment by apaszke** [comment]
@davidbau that's a great link! The comment at the beginning of that file seems to outline the whole algorithm and looks pretty solid.

**Comment by apaszke** [comment]
Yeah the paper "Use of Extended Euclidean Algorithm in Solving a System of Linear Diophantine Equations with Bounded Variables" pretty much summarises what I've been thinking of, except it corrects a mistake in my quick analysis that made the algorithm to appear to be linear in `d` instead of exponential 😕

Considering this, I think that if we want to handle those cases, then we should implement a few heuristics and forget about solving the general equations (just always fall back to a warning). It's both expensive and hard to implement (due to integer overflows).

**Comment by t-vi** [comment]
So the heuristics would cover the typical cases outlined in the [numpy doc comment](https://github.com/numpy/numpy/blob/9a4d75dce2442b81859152048c299c57e6610667/numpy/core/src/private/mem_overlap.c#L80-L112)?

**Comment by apaszke** [comment]
I think that would be reasonable.

**Comment by t-vi** [comment]
So I looked somewhat further (unfortunately, I first did the easy "surely OK" cases that aren't that useful by themselves). 
One thing that I wondered about: Given that GCC and Clang support int128, would you accept a full implementation that uses those (MSVC doesn't support int128, so they would miss out without further dependencies)?

**Comment by apaszke** [comment]
Why would we need int128 though? We never want to go down the path of solving those equations I think, we just want the rough common-case heuristics

**Comment by t-vi** [comment]
OK. I wanted to solve them. But then I'll restrain myself.

**Comment by fmassa** [comment]
I'm raising the priority of this to high, given that this issue keeps coming over and over again. I've created a `topic: partial aliasing` label to simplify searching for those issues https://github.com/pytorch/pytorch/issues?q=is%3Aissue+label%3A%22topic%3A+partial+aliasing%22+is%3Aopen

**Comment by ezyang** [comment]
@pbelevich is working on this
